posthog/temporal/data_imports/row_tracking.py

```diff
@@ -76,3 +83,95 @@ def get_rows(team_id: int, schema_id: uuid.UUID | str) -> int:
                 return int(value)
 
         return 0
+
+
+def get_all_rows_for_team(team_id: int) -> int:
+    with _get_redis() as redis:
+        if not redis:
+            return 0
+
+        pairs = redis.hgetall(_get_hash_key(team_id))
+        return sum(int(v) for v in pairs.values())
+
+
+def will_hit_billing_limit(team_id: int, logger: FilteringBoundLogger) -> bool:
+    if not EE_AVAILABLE:
+        return False
+
+    try:
+        from ee.billing.billing_manager import BillingManager
+
+        logger.debug("Running will_hit_billing_limit")
+
+        license = get_cached_instance_license()
+        billing_manager = BillingManager(license)
+        team = Team.objects.get(id=team_id)
+        organization: Organization = team.organization
+        all_teams_in_org: list[int] = [
+            value[0] for value in Team.objects.filter(organization_id=organization.id).values_list("id")
+        ]
+
+        logger.debug(f"will_hit_billing_limit: Organisation_id = {organization.id}")
+        logger.debug(f"will_hit_billing_limit: Teams in org: {all_teams_in_org}")
+
+        billing_res = billing_manager.get_billing(organization)
+
+        logger.debug(f"will_hit_billing_limit: billing_res = {billing_res}")
+
+        current_billing_cycle_start = billing_res.get("billing_period", {}).get("current_period_start")
+        if current_billing_cycle_start is None:
+            logger.debug(
+                f"will_hit_billing_limit: returning early, no current_period_start available. current_billing_cycle_start = {current_billing_cycle_start}"
+            )
+            return False
+
+        current_billing_cycle_start_dt = parser.parse(current_billing_cycle_start)
+
+        logger.debug(f"will_hit_billing_limit: current_billing_cycle_start = {current_billing_cycle_start}")
+
+        usage_summary = billing_res["usage_summary"]
+        rows_synced_summary = usage_summary.get("rows_synced", None)
+
+        if not rows_synced_summary:
+            logger.debug(
+                f"will_hit_billing_limit: returning early, no rows_synced key in usage_summary. {usage_summary}"
+            )
+            return False
+
+        rows_synced_limit = rows_synced_summary.get("limit")
+
+        logger.debug(f"will_hit_billing_limit: rows_synced_limit = {rows_synced_limit}")
+
+        if rows_synced_limit is None or not isinstance(rows_synced_limit, int | float):
+            logger.debug("will_hit_billing_limit: rows_synced_limit is None or not a number, returning False")
+            return False
+
+        # Get all completed rows for all teams in org
+        rows_synced_in_billing_period_dict = ExternalDataJob.objects.filter(
+            team_id__in=all_teams_in_org,
+            finished_at__gte=current_billing_cycle_start_dt,
```

Comment:
I don't think i'm following the conversation above. Why don't we filter for all the jobs ahead of the latest usage report? So historical is billing data and latest is from the jobs directly

Reply:
There's a time gap between the usage reports being updated for "yesterday" and when "today" starts. e.g. when midnight strikes, the data from billing service won't be fully updated, nor do we get a timestamp for when the metrics were updated.

**Example:**
It's 02:00am UTC on **June 5th**, billing-service will report that there have been 120,000 rows synced up until "yesterday", but "yesterday" could mean June 3rd or June 4th - it depends on whether the cron job for updating this has completed yet, and so we could be undercounting the rows by a bunch when checking